server/monitor.go

```diff
@@ -0,0 +1,504 @@
+	tsecs := d / time.Second
+	tmins := tsecs / 60
+	thrs := tmins / 60
+	tdays := thrs / 24
+	tyrs := tdays / 365
+
+	if tyrs > 0 {
+		return fmt.Sprintf("%dy%dd%dh%dm%ds", tyrs, tdays%365, thrs%24, tmins%60, tsecs%60)
+	}
+	if tdays > 0 {
+		return fmt.Sprintf("%dd%dh%dm%ds", tdays, thrs%24, tmins%60, tsecs%60)
+	}
+	if thrs > 0 {
+		return fmt.Sprintf("%dh%dm%ds", thrs, tmins%60, tsecs%60)
+	}
+	if tmins > 0 {
+		return fmt.Sprintf("%dm%ds", tmins, tsecs%60)
+	}
+	return fmt.Sprintf("%ds", tsecs)
+}
+
+func (s *StanServer) handleStorez(w http.ResponseWriter, r *http.Request) {
+	count, bytes, _ := s.store.MsgsState(stores.AllChannels)
+	storez := &Storez{
+		ClusterID:  s.info.ClusterID,
+		ServerID:   s.serverID,
+		Now:        time.Now(),
+		Type:       s.store.Name(),
+		Limits:     *s.storeLimits,
+		TotalMsgs:  count,
+		TotalBytes: bytes,
+	}
+	sendResponse(w, r, storez)
+}
+
+type byClientID []*Clientz
+
+func (c byClientID) Len() int           { return len(c) }
+func (c byClientID) Swap(i, j int)      { c[i], c[j] = c[j], c[i] }
+func (c byClientID) Less(i, j int) bool { return c[i].ID < c[j].ID }
+
+func (s *StanServer) handleClientsz(w http.ResponseWriter, r *http.Request) {
+	singleClient := r.URL.Query().Get("client")
+	subsOption, _ := strconv.Atoi(r.URL.Query().Get("subs"))
+	if singleClient != "" {
+		var clientz *Clientz
+		client := s.store.GetClient(singleClient)
+		if client != nil {
+			clientz = getMonitorClient(client, subsOption)
+		}
+		if clientz == nil {
+			w.WriteHeader(http.StatusNotFound)
+			w.Write([]byte(fmt.Sprintf("Client %s not found", singleClient)))
+			return
+		}
+		sendResponse(w, r, clientz)
+	} else {
+		offset, _ := strconv.Atoi(r.URL.Query().Get("offset"))
+		if offset < 0 {
+			offset = 0
+		}
+		limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
+		if limit <= 0 {
+			limit = defaultMonitorListLimit
+		}
+		clients := s.store.GetClients()
+		totalClients := len(clients)
+		carr := make([]*Clientz, 0, totalClients)
+		for _, c := range clients {
+			cz := &Clientz{ID: c.ID}
+			carr = append(carr, cz)
+		}
+		sort.Sort(byClientID(carr))
+
+		minoff := offset
+		if minoff > totalClients {
+			minoff = totalClients
+		}
+		maxoff := offset + limit
+		if maxoff > totalClients {
+			maxoff = totalClients
+		}
+		carr = carr[minoff:maxoff]
+
+		for _, c := range carr {
+			cli := clients[c.ID]
+			c.HBInbox = cli.HbInbox
+			if subsOption == 1 {
+				srvCli := cli.UserData.(*client)
+				c.Subscriptions = getMonitorClientSubs(srvCli, true)
+			}
+		}
+		clientsz := &Clientsz{
+			ClusterID: s.info.ClusterID,
+			ServerID:  s.serverID,
+			Now:       time.Now(),
+			Offset:    offset,
+			Limit:     limit,
+			Total:     totalClients,
+			Count:     len(carr),
+			Clients:   carr,
+		}
+		sendResponse(w, r, clientsz)
+	}
+}
+
+func getMonitorClient(c *stores.Client, subsOption int) *Clientz {
+	cli := c.UserData.(*client)
+	cli.RLock()
+	if cli.unregistered {
+		cli.RUnlock()
```

Comment:
Sure, there should not be million of clients (because this is call for every client in the list of clients returned by the `/clientsz` endpoint).

As a general rule, though, I don't use defer if it is fast path, and if the code complexity is so low that it is easy to not miss the unlock. For instance, for a function that would do something as simple as returning an object's field, I would prefer this to a defer:
```
func (s *mystruct) getObj() *obj {
  s.Lock()
  o := s.obj
  s.Unlock()
  return o
}
```  
(I may have written some code that does just the opposite, but the idea is that it's not because the language provides defer that you have to use it all the times. And again, if that function was called only once in a while, it does not matter, but too often, and I would really prefer not to use the defer).